[PBA-5493] Show Discovery on end screen correctly

brol1dev opened the pull request:
We weren’t showing the discovery on the end screen with showUpNext = false and screenToShowOnEnd = “discovery”. This was a bug since we would only show discovery on end screen if we dismissed the upNext widget previously. Added a logic operation bypassing the upNextDismissed question if the showUpNext button was set to false, this way we would only care about screenToShowOnEnd.

I wanted to add testing to this change but I found out that with our current dependencies versions testing is broken. We will have to fix it first and then we should add more testing to our react native code to find this kind of errors.

Commit:
[PBA-5493] Show Discovery on end screen correctly

We weren’t showing the discovery on the end screen with showUpNext = false and screenToShowOnEnd = “discovery”. This was a bug since we would only show discovery on end screen if we dismissed the upNext widget previously. Added a logic operation bypassing the upNextDismissed question if the showUpNext button was set to false, this way we would only care about screenToShowOnEnd.

brol1dev commented:
Remember we need two approvals before merging in the changes.

akshaydch11 commented:
Looks good to me 👍

aeng7 commented on sdk/react/ooyalaSkinCore.js:
```diff
@@ -133,7 +133,7 @@ OoyalaSkinCore.prototype.shouldShowLandscape = function() {
 
 OoyalaSkinCore.prototype.shouldShowDiscoveryEndscreen = function() {
-  return (this.skin.state.upNextDismissed == true && this.skin.props.endScreen.screenToShowOnEnd == "discovery");
+  return (!this.skin.props.upNext.showUpNext || this.skin.state.upNextDismissed) && this.skin.props.endScreen.screenToShowOnEnd === "discovery";
```
Should we split up this line into multiple lines for readability?

brol1dev replied:
sure

Commit:
[PBA-5493] Use more readable code

brol1dev commented:
Updated code, @aeng7. I'm very used to use as little code as possible, sorry about that :P

brol1dev commented:
Oh wait, i'm going to change something else, don't merge yet!

Commit:
[PBA-5493] Deal with boolean values only for showUpNext setting

Being cautious about how we deal with the settings provided in skin.json, showUpNext should be a bool value but the customer can provide anything they want. Right now if they provide something different than a boolean strange behavior may occur with showing the upNext widget or not showing it, and also showing the discovery on end screen. If customers use the skin.json as it should be they won’t encounter any problems.

brol1dev commented:
@akshaydch11, @aeng, review once more please, since I added some more changes in other impacted parts.

akshaydch11 commented:
Looks Good 👍

jpfranco commented on sdk/react/ooyalaSkinCore.js:
```diff
@@ -133,7 +133,24 @@ OoyalaSkinCore.prototype.shouldShowLandscape = function() {
 OoyalaSkinCore.prototype.shouldShowDiscoveryEndscreen = function() {
-  return (this.skin.state.upNextDismissed == true && this.skin.props.endScreen.screenToShowOnEnd == "discovery");
+  // Only care if discovery on endScreen should be shown
+  if (this.skin.props.endScreen.screenToShowOnEnd !== "discovery") {
```
Do we populate this config with default values if, for example, `endScreen` is not provided?  I'm not sure if this would be an issue on the PBA side, but on the html5 skin we started using a safe get function so that this wouldn't crash if the skin config was incomplete.

brol1dev replied:
pushed a change for what you're saying. This is why we need testing in place. We'll have to work towards it little by little. Thanks!

Commit:
[PBA-5493] Checking if we don’t have a valid config

If we remove the endScreen and/or upNext configs in skin.json crashes were occurring in multiple parts. I added a safety check to see if we have valid config objects, if not we’ll create an empty one which is safer than having undefined values.

brol1dev commented:
@jpfranco please review once more when you have a chance

brol1dev commented:
I still need a second approval. Friendly reminder @jpfranco, @dchauOoyala, @aeng7. Ask me anything you need to know. 

Thanks :)

brol1dev commented:
guys, if no one comments today after our sprint planning I will merge it myself with one approval only. This PR is required for the upcoming release.

dchauOoyala commented on sdk/react/panels/EndScreen.js:
```diff
@@ -69,10 +69,11 @@ var EndScreen = React.createClass({
   _renderDefaultScreen: function(progressBar, controlBar) {
-    var fullscreenPromoImage = (this.props.config.endScreen.mode == 'default');
+    var endScreenConfig = this.props.config.endScreen || {};
+    var fullscreenPromoImage = endScreenConfig === 'default';
```
I notice the interchanging of single `''` and double `""` quotes throughout. Is there a reason for this?

brol1dev replied:
We should be using single quotes, but we have double quotes all over the place. I'll change the ones I see in the files I'm updating currently.

brol1dev replied:
Oh btw, single quotes only apply to JS. Java and ObjC must use double quotes.

brol1dev replied:
omg, we don't even use this fullscreenPromoImage. I'm just deleting this line.

dchauOoyala commented on sdk/react/panels/EndScreen.js:
```diff
@@ -69,10 +69,11 @@ var EndScreen = React.createClass({
 
   _renderDefaultScreen: function(progressBar, controlBar) {
-    var fullscreenPromoImage = (this.props.config.endScreen.mode == 'default');
+    var endScreenConfig = this.props.config.endScreen || {};
```
Is `this.props.config.endScreen` sometimes `String` and sometimes an object?

brol1dev replied:
it is always an object, and thx for catching the error `endScreenConfig === 'default'` :)

Commit:
[PBA-5493] Be consistent with quotes usage in string literals

Use double quotes everywhere moving forward. This is consistent with PBW practices.

brol1dev commented:
done @dchauOoyala, please review once more or merge if it's ok :)

dchauOoyala commented:
In regards to the quotes, there is not a mandated practice of using one over the other. It comes down to preference, but I personally prefer double quotes over single to remain consistent across languages. Web code currently leans towards double quotes. However, consistency is key so whichever one we do decide to use here for PBA we should do so moving forward.

dchauOoyala commented:
LGTM
